File: android-app/app/src/main/java/app/t4l/domain/DeterministicIds.kt
package app.t4l.domain

import java.nio.ByteBuffer
import java.security.MessageDigest
import java.util.UUID

object DeterministicIds {
    fun budgetAllocation(workspaceId: String, planId: String, categoryId: String): String {
        val name = "budget-allocation:${workspaceId.lowercase()}:${planId.lowercase()}:${categoryId.lowercase()}"
        val bytes = MessageDigest.getInstance("SHA-256").digest(name.toByteArray(Charsets.UTF_8)).copyOf(16)
        bytes[6] = ((bytes[6].toInt() and 0x0f) or 0x50).toByte()
        bytes[8] = ((bytes[8].toInt() and 0x3f) or 0x80).toByte()
        val buffer = ByteBuffer.wrap(bytes)
        return UUID(buffer.long, buffer.long).toString()
    }
}

File: server/src/T4L.Api/Sync/DeterministicIds.cs
using System.Security.Cryptography;
using System.Text;

namespace T4L.Api.Sync;

public static class DeterministicIds
{
    public static Guid Named(string name)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(name))[..16];
        bytes[6] = (byte)((bytes[6] & 0x0f) | 0x50);
        bytes[8] = (byte)((bytes[8] & 0x3f) | 0x80);
        return new Guid(bytes, bigEndian: true);
    }

    public static Guid BudgetAllocation(Guid workspaceId, Guid planId, Guid categoryId)
        => Named($"budget-allocation:{workspaceId:D}:{planId:D}:{categoryId:D}".ToLowerInvariant());
}
